woocommerce/google-listings-and-ads
# Cleanup synced products locally on MC disconnect

### Changes proposed in this Pull Request:

This PR resolves the scenario where a Merchant may disconnect their account and then reconnect another account. When they do this all the products are still marked as synced locally until the new product sync has completed. This PR creates a cleanup job that runs when the Merchant Center is disconnected and will mark all products as unsynced. Since it can take a while to complete, I added a check to only continue unsyncing the products as long as the Merchant Center account remains disconnected. Once a new account it reconnected it will stop, any remaining products will later be cleaned up when a new Product Sync happens.

Closes #1233 

### Detailed test instructions:

I added unit testing for the `CleanupSyncedProducts` job, as well as the added `BatchProductHelper` function. However there are no unit tests present for disconnecting the Merchant Center, it was added in the pending PR #1226, once that is merged we can create a followup to confirm the scheduled cleanup is triggered on disconnect.

Steps to test
1. Connect a Merchant account and Sync all products to Google
2. Disconnect the Merchant account, and wait a while till scheduled actions have been run
3. Go to Tools > Scheduled Actions and search for `cleanup_synced`
![image](https://user-images.githubusercontent.com/11388669/152953043-60dcf9d0-b93c-4ebd-9cac-5f0ebd5bd9ca.png)
6. Confirm you see all the product IDs that were previously synced (could be spread out across multiple scheduled actions if there is a large amount)

### Changelog entry
* Fix - Cleanup synced products locally when disconnecting Merchant Center account.
